### PR TITLE
fix: compute kd loss in trainer to bypass broken patch/inject paths

### PR DESCRIPTION
**Pull request by roycho96**
# Description

Two coupled bugs in the KD trainer made KD training crash or silently train wrong on current axolotl (transformers 5.5.4).

1. `AxolotlKDTrainer.__init__` injected `LigerFusedLinearKLTopKLogprobLoss` into `model._loss_function`. Its signature is `(lm_head_weight, hidden_states, target_token_ids, ...)`. Transformers 5.x calls `self.loss_function(logits, labels, vocab_size, ...)` from the stock forward path. The two signatures do not match, so any code path that does not go through the patched forward crashes with a TypeError.
2. `apply_kernel(cfg.model_config_type)` in `pre_model_load` patched `{prefix}ForCausalLM.forward`. For multimodal models (Gemma3, Gemma4) the actually instantiated class is `XxxForConditionalGeneration`, not `XxxForCausalLM`. The patch hits the wrong class and silently does nothing, so the model falls back to the stock forward and triggers bug 1.

This PR drops both mechanisms and computes the KD loss inside `compute_loss` in the trainer.

## Motivation and Context

KD training on transformers 5.x or on multimodal Gemma was broken. Either it crashed immediately with a TypeError, or it appeared to start but used the wrong forward path depending on which class got patched.

The fix removes the model-side hooks entirely. `compute_loss` requests `output_hidden_states=True` from the model, picks the last hidden state, finds the `lm_head` on the model itself (with PEFT unwrap and multimodal `language_model.lm_head` fallback), and calls the KD loss inside the trainer. The model classes and instances are untouched, so any model family works as long as it exposes `lm_head` or `language_model.lm_head`.

Side effect: `kernels/models.py` (which defined the old `apply_kernel` and `kldiv_forward_llama_like`) is no longer reachable and is removed in a follow-up commit on this PR.

Known limitation (not addressed here): when LoRA is applied directly to `lm_head`, `lm_head.weight` returns the unmerged base weight. Most KD recipes do not put LoRA on `lm_head`, so this is left as a documented constraint.

## How has this been tested?

12 unit tests in `tests/integrations/test_kd_trainer_direct_loss.py`:

- `_resolve_lm_head` resolution order: plain CausalLM, multimodal `language_model.lm_head`, PEFT wrapper, PEFT wrapper around multimodal, and the missing-lm_head error case.
- `compute_loss` happy path: asserts the KD loss is called with the right tensor shapes and the `true_labels` kwarg, and that none of the KD-only keys (`labels`, `target_token_ids`, `target_logprobs`, `target_mask`, `num_items_in_batch`) leak into the model forward.
- `num_items_in_batch` handling: computed from non-ignored labels, overridden when passed explicitly, and skipped when zero.
- Error paths: missing KD keys raise a clear `KeyError`; missing `hidden_states` raises a clear `RuntimeError`.
- Input dict immutability: the caller's `inputs` dict is not mutated.

## AI Usage Disclaimer

Opus was used to draft the regression tests and review the diff.

## Types of changes

- [x] Bug fix

## Summary by CodeRabbit

* **Refactor**
  * Restructured Knowledge Distillation training pipeline with updated loss computation approach
  * Removed pre-model-loading initialization hooks and consolidated related logic

* **Tests**
  * Added comprehensive integration test suite for KD trainer functionality, including loss computation and model resolution verification

[Review Change Stack](https://app.coderabbit.ai/change-stack/axolotl-ai-cloud/axolotl/pull/3661?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/axolotl/integrations/kd/trainer.py [hunk 1]
```diff
@@ -70,34 +89,49 @@ def compute_loss(
...
 
+        inputs["output_hidden_states"] = True
         outputs = model(**inputs)
+        hidden_states = getattr(outputs, "hidden_states", None)
+        if hidden_states is None:
+            raise RuntimeError(
+                f"{type(model).__name__} did not return hidden_states despite "
+                "output_hidden_states=True; AxolotlKDTrainer requires hidden state access."
+            )
+        hidden_states = hidden_states[-1]
```
[the remark this answers is not in the document]
Comment (roycho96) [reply src/axolotl/integrations/kd/trainer.py:104]: done

**Review by winglian — APPROVED** [review]
